packages/engine/paima-sm/src/delegate-wallet.ts

```diff
@@ -0,0 +1,307 @@
+  getAddressFromAddress,
+  getDelegation,
+  getDelegationsTo,
+  getDelegationsToWithAddress,
+  newAddress,
+  newDelegation,
+  updateAddress,
+} from '@paima/db';
+
+type AddressExists = { address: IGetAddressFromAddressResult; exists: true };
+type AddressDoesNotExist = { address: string; exists: false };
+type AddressWrapper = AddressExists | AddressDoesNotExist;
+
+type ParsedDelegateWalletCommand =
+  | {
+      command: 'delegate';
+      args: { from: string; to: string; from_signature: string; to_signature: string };
+    }
+  | {
+      command: 'migrate';
+      args: { from: string; to: string; from_signature: string; to_signature: string };
+    }
+  | { command: 'cancelDelegations'; args: { to: string; to_signature: string } };
+
+// Delegate Wallet manages cache cleanup.
+enableManualCache();
+
+export class DelegateWallet {
+  private static readonly DELEGATE_WALLET_PREFIX = 'DELEGATE-WALLET';
+
+  public static readonly INTERNAL_COMMAND_PREFIX = '&';
+
+  private static readonly SEP = ':';
+
+  private static readonly delegationGrammar = `
+    delegate            =   &wd|from|to|from_signature|to_signature
+    migrate             =   &wm|from|to|from_signature|to_signature
+    cancelDelegations   =   &wc|to|to_signature
+    `;
+
+  private static readonly parserCommands = {
+    delegate: {
+      from: PaimaParser.WalletAddress(),
+      to: PaimaParser.WalletAddress(),
+      from_signature: PaimaParser.NCharsParser(0, 1024),
+      to_signature: PaimaParser.NCharsParser(0, 1024),
+    },
+    migrate: {
+      from: PaimaParser.WalletAddress(),
+      to: PaimaParser.WalletAddress(),
+      from_signature: PaimaParser.NCharsParser(0, 1024),
+      to_signature: PaimaParser.NCharsParser(0, 1024),
+    },
+    cancelDelegations: {
+      to: PaimaParser.WalletAddress(),
+      to_signature: PaimaParser.NCharsParser(0, 1024),
+    },
+  };
+
+  private static parser = new PaimaParser(
+    DelegateWallet.delegationGrammar,
+    DelegateWallet.parserCommands
+  );
+
+  constructor(private DBConn: PoolClient) {}
+
+  /* Generate Plaintext Message */
+  private generateMessage(internalMessage: string = ''): string {
+    return `${DelegateWallet.DELEGATE_WALLET_PREFIX}${
+      DelegateWallet.SEP
+    }${internalMessage.toLocaleLowerCase()}${DelegateWallet.SEP}${ENV.CONTRACT_ADDRESS}`;
+  }
+
+  /* Verify Signature with all possible wallets. */
+  private async verifySignature(
+    walletAddress: string,
+    message: string,
+    signature: string
+  ): Promise<boolean> {
+    if (!walletAddress || !signature) throw new Error('No Signature');
+    const wallets = [
+      CryptoManager.Evm(new Web3()),
+      CryptoManager.Algorand(),
+      CryptoManager.Cardano(),
+      CryptoManager.Polkadot(),
+    ];
+    for (const wallet of wallets) {
+      try {
+        if (await wallet.verifySignature(walletAddress, message, signature)) {
+          return true;
+        }
+      } catch {
+        // do nothing.
+      }
```

Comment:
This is going to run really slowly (esp. if we add more currencies). Probably we want to detect which algorithm to use in some other way (either by checking which address submitted this or through an extra field)

Reply:
By address (would this work?)

eth: `'0x[a-fA-F0-9]{40}'` 
cardano `addr1[a-z0-9]+` 
polkadot `[1-9A-HJ-NP-Za-km-z]{47,48}` https://namespaces.chainagnostic.org/polkadot/caip10
algorand ` [A-Z2-7]{58}` 

We would have to add 2 params to wallet delegation 
`&wd|from|to|from_signature|to|signature|from_type|to_type`

What path do we take?

Reply:
I don't think we need to add anything to the concise encoding:
1. If the user uses the batcher, we already know the address since the batch is of the form `userAddress/userSignature/input/millisecondTimestamp`
2. If the user self-sequences, we know the address since they're the ones who submitted the transaction